### PR TITLE
Add a size comparison between futures-based async code and non-futures-based async code.

### PR DESCRIPTION
**Pull request by jrvanwhy**
This size comparison evaluates the code size impact of adding futures to `libtock-rs` (and potentially the Tock kernel in the future).

### CODE REVIEW

**Comment by jrvanwhy** [comment]
@Woyten and @torfmaster: I'd like you to take a look as well. I'm not sure how to add you as reviewers directly.

File: size_comparison/.cargo/config [hunk 1]
```diff
@@ -0,0 +1,8 @@
+[build]
+target = "thumbv7m-none-eabi"
```
Comment (torfmaster) [inline size_comparison/.cargo/config:2]: Isn't the majority of boards we're supporting cortex-m4?

File: size_comparison/build.sh [hunk 2]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+cargo build --release
```
Comment (torfmaster) [inline size_comparison/build.sh:3]: you should specify the target here and `--workspace` is also required.

**Comment by torfmaster** [comment]
I have some comments to your work (which I will post separately). First of all I built a very tiny cli which extracts the desired numbers and [prints them](https://github.com/torfmaster/libtock-size-utils). Whenever we are talking about numbers we should make them available in the tool (and eventually write tests which can assert on the numbers). There is a difference of 4 bytes to your numbers but I think it should be neglectible (the offset also occurs when I use readelf) it may be that the build is not 100% deterministic.

**Comment by torfmaster** [comment]
One aspect is that the `Future` implementation is quite different from what we currently use in `libtock-rs`. We currently don't use a waker implementation at all. I will take me some time to analyze your approach to futures but maybe you can give some motivation about your approach (and maybe even create a PR to `libtock-rs` to change the current approach for futures if this is benefitial).

**Comment by torfmaster** [comment]
I used the size tooling to do some very rough investigation of sizes. I will try to come up with more complex examples but implementing complex logic with the callback approach takes some time.

I just post the numbers here for reference:
Also, the numbers are quite different for the examples (which are quite unspecific to our size investigation, I know).

Example blink: current master (08382104152f4e95e91307483202ceba0ca1fc4d):
Size of sections in bytes: 
        text: 3028 
        data: 0
        rodata: 0
        bss: 16

Example blink (ab2c945184b98ecae3e70ac678e9f5231deef73b)
Size of sections in bytes: 
        text: 2796 
        data: 0
        rodata: 0
        bss: 20

Same for `blink_random` (which has more complicated structure):
Async version:
Size of sections in bytes: 
        text: 4080 
        data: 0
        rodata: 0
        bss: 16

Sync version:
Size of sections in bytes: 
        text: 3728 
        data: 0
        rodata: 0
        bss: 20

I currently cannot reproduce the high cost in terms of text segment size the current penality seems to be around 5%.

File: size_comparison/.cargo/config (hunk 1 above)
Comment (jrvanwhy) [reply size_comparison/.cargo/config:2]: Perhaps, but thumbv7m-none-eabi is what I have available. The particular architecture doesn't really matter here.

File: size_comparison/build.sh (hunk 2 above)
Comment (jrvanwhy) [reply size_comparison/build.sh:3]: What is `--workspace`? I don't see its documentation. Why should I specify the target here? It's pulled from `.config/cargo`.

**Comment by jrvanwhy** [comment]
> One aspect is that the `Future` implementation is quite different from what we currently use in `libtock-rs`. We currently don't use a waker implementation at all. I will take me some time to analyze your approach to futures but maybe you can give some motivation about your approach (and maybe even create a PR to `libtock-rs` to change the current approach for futures if this is benefitial).

I don't see how using `core::future::Future` without using wakers is compatible with the futures ecosystem. The primary benefit of futures is interoperability, so we should do this in a compatible manner.

Does `libtock-rs`'s current implementation just poll everything every time a callback runs?

**Comment by torfmaster** [comment]
> I don't see how using core::future::Future without using wakers is compatible with the futures ecosystem. The primary benefit of futures is interoperability, so we should do this in a compatible manner.

I think the import question is whether a trivial waker implementation can break compatibilty. I currently see no obvious case where `libtock`'s waker violates the contracts for futures or even causes problems but of course this is no proof. So the important question would be: Can we run arbitrary futures on our executor. An invalid question in my opinion would be whether we can run all of our futures on arbitrary executors.

> Does libtock-rs's current implementation just poll everything every time a callback runs?

`block_on` is currently polling the future (only) it is given every time the app wakes up from a yield. Using `async`/`.await` the number of futures polled is bounded by the number of pending futures (at least when I understand the state machines which are generated correctly). This sounds like a worst case scenario for concurrency, however I expect the number of this kind of futures to be quite low.

File: size_comparison/.cargo/config (hunk 1 above)
Comment (torfmaster) [reply size_comparison/.cargo/config:2]: I see

File: size_comparison/build.sh (hunk 2 above)
Comment (torfmaster) [reply size_comparison/build.sh:3]: You're right. I actually broke the configuration when switching the target for me.

**Comment by jrvanwhy** [comment]
> > I don't see how using core::future::Future without using wakers is compatible with the futures ecosystem. The primary benefit of futures is interoperability, so we should do this in a compatible manner.
> 
> I think the import question is whether a trivial waker implementation can break compatibilty. I currently see no obvious case where `libtock`'s waker violates the contracts for futures or even causes problems but of course this is no proof. So the important question would be: Can we run arbitrary futures on our executor. An invalid question in my opinion would be whether we can run all of our futures on arbitrary executors.

I was under the impression that futures could create their own wakers and expect them to be called, but I don't see any evidence of that any more. Perhaps that changed in future 0.3? I'll put together a version with a trivial waker and see whath the impact is.

The downside of polling everything all the time is that each callback will result in work proportional to the total application size. I have no idea whether that could be a significant limitation. It also means that you an app developer needs to bring all futures back to a single point (i.e. `main()`) or we need a list of tasks to poll.

**Comment by torfmaster** [comment]
>  I'll put together a version with a trivial waker and see whath the impact is.

Wouldn't it be better to refer to the actual `libtock-rs`. I know you want to isolate the problem (which is interesting) but the benefit would probably be higher if we actually _used_ libtock.

> The downside of polling everything all the time is that each callback will result in work proportional to the total application size.

I'm not sure about that. I think the work (say total polls per unit of time) will be bounded by something KxN where K is the number of concurrent pending futures (a number which can be computed at compile time when you have no heap) and N is the number of concurrent active callbacks firing per unit of time.

> It also means that you an app developer needs to bring all futures back to a single point (i.e. main()) or we need a list of tasks to poll.

I think futures are easy to compose using either sequential execution (`.await`) or parallel execution (`join`) so this is not a real limitation.

**Comment by jrvanwhy** [comment]
> Wouldn't it be better to refer to the actual `libtock-rs`. I know you want to isolate the problem (which is interesting) but the benefit would probably be higher if we actually _used_ libtock.

What would be the benefit of using the real `libtock`? In order to evaluate the impact of futures, I need reasonably-well-optimized examples. Doing that with the real `libtock` is a lot of effort that may be completely wasted.

**Comment by torfmaster** [comment]
> What would be the benefit of using the real libtock? In order to evaluate the impact of futures, I need reasonably-well-optimized examples. Doing that with the real libtock is a lot of effort that may be completely wasted.

My idea would be to evaluate the way how we _actually_ use futures to help the potential user to find out whether async-await is something she wants or not. But anyway, it depends on want you want to measure. Measuring just sizes of programs which do the same thing is easy. If you want to do more sophisticated things (as identifying now-mangled functions in the binary) it may become more compilcated.

**Comment by jrvanwhy** [comment]
> My idea would be to evaluate the way how we _actually_ use futures to help the potential user to find out whether async-await is something she wants or not. But anyway, it depends on want you want to measure. Measuring just sizes of programs which do the same thing is easy. If you want to do more sophisticated things (as identifying now-mangled functions in the binary) it may become more compilcated.

I'm trying to identify whether `libtock-rs` needs to expose a non-futures-based async API or whether it's fine for it to only support a futures-based API.

**Comment by torfmaster** [comment]
> I'm trying to identify whether libtock-rs needs to expose a non-futures-based async API or whether it's fine for it to only support a futures-based API.

But then I don't understand why you take an alternative implementation of the futures api? Do you want to have a more minimal/complete version of it? Or do you think there's too much noise coming from the rest of the library and hence create a minimal implementation. But I see we're trying to answer different questions here.

**Comment by jrvanwhy** [comment]
> Or do you think there's too much noise coming from the rest of the library and hence create a minimal implementation.

I'm trying to remove artifacts of `libtock-rs`'s current implementation from the comparison. I have found it to be quite noisy, and it is far easier to implement my own than to remove that noise.

**Comment by torfmaster** [comment]
I have some other questions (mostly about the requirement of having a small application size)

 * How big are the application compared to the kernel size?
 * Are size limits actually hit. Do we support hardware where we could hit the limits (the nrf52dk has ~640 kbytes of flash so it is not a real issue _there_)?
 * How big are applications compared to `libtock-c`?
 * What is the expected number of concurrent applications running on one kernel?
 * What is the performance penalty (e.g. for not monomorphizing calls)?

**Comment by jrvanwhy** [comment]
> I have some other questions (mostly about the requirement of having a small application size)
> 
> * How big are the application compared to the kernel size?

For H1B, the kernel is 132 KiB in flash and the applications are limited to 32 KiB (practically speaking, the kernel + apps are limited to 256KiB, but there are MPU limits as well that prevent that space from being filled). The kernel is currently allocated 16 KiB of RAM and the apps are allocated 48 KiB.

> * Are size limits actually hit. Do we support hardware where we could hit the limits (the nrf52dk has ~640 kbytes of flash so it is not a real issue _there_)?

The existing (non-Tock-based) firmware on Google's H1B chip is space-constrained. Moving to Tock makes the space problem larger because we need to add the overhead of the system call API interfaces and redundant code. I don't know if we've hit RAM limits.

> * How big are applications compared to `libtock-c`?

I don't know, perhaps someone else can comment. I'm not sure how easy or hard this is to measure, given that link time optimization is in use.

> * What is the expected number of concurrent applications running on one kernel?

At least 4.

> * What is the performance penalty (e.g. for not monomorphizing calls)?

Are you referring to the CPU time overhead of calling through vtables? If so, probably small, as the platforms Tock runs on generally aren't superscalar (or don't pipeline many instructions simultaneously if they are).

**Comment by phil-levis** [comment]
> I have some other questions (mostly about the requirement of having a small application size)
> 
>     * How big are the application compared to the kernel size?

This is of course system and application dependent. There are cases when you want a lean kernel and a lot of code in userspace. Others when you want a big kernel and thin applications. Because embedded systems are by definition application-specific systems, we need to support both cases.

> 
>     * Are size limits actually hit. Do we support hardware where we could hit the limits (the nrf52dk has ~640 kbytes of flash so it is not a real issue _there_)?

Yes they are. Just because one platform has more code space than it needs doesn't mean all do. Some of Google's Titan use cases absolutely hit the code size ceiling. To put your question in a different light. "Are CPUs actually run at 100%? My laptop never hits 100% except if there is a bad web page." Historically, we always hit the code limit. The one caveat in embedded systems is that you can sometimes hit RAM before code: a 1:10 ratio is usual rule of thumb.


>     * What is the expected number of concurrent applications running on one kernel?

Application-specific. The basic point is that we don't want to say "Eh, 4 is enough, we'll never need more than 4" because then of course someone will need 5. E.g., recall when browsers shifted to putting each tab in a separate process for memory isolation?
 
>     * What is the performance penalty (e.g. for not monomorphizing calls)?

Great question -- I think we can evaluate this experimentally.

**Comment by Woyten** [comment]
> The downside of polling everything all the time is that each callback will result in work proportional to the total application size. I have no idea whether that could be a significant limitation. It also means that you an app developer needs to bring all futures back to a single point (i.e. `main()`) or we need a list of tasks to poll.

The current situation of `libtock-rs` is that you do *not* poll everything. You only poll *the* future of the `main` function. Considering that Tock processes are single-threaded and assuming that yielding within callbacks is bad, a single main future seems sufficient. Thus, there is no need for a waker keeping track which future has been awoken. To prevent multiple futures being spawned `block_on` is defined to be `unsafe`.

Using the `join` function it is possible to schedule multiple futures concurrently. It is true that this means every that subsequent future is `poll`ed each time the outer `JoinFuture` is `poll`ed. This is an implementation detail of the `join` function which is out of our reach. I am not sure whether a `JoinFuture` is directly comparable to two simultaneously subscribed callbacks but the more I think about it, it might be clever to offer both a synchronous and an asynchronous API.

Regarding the point you mentioned about each callback resulting in some work: This has always been the case when using `yieldk_for`.

**Comment by Woyten** [comment]
I did some measurements as well. It turned out that the `text` section is growing with the commit that introduced the futures but there are plenty of other commits which compensate for this effect. The size increase has been different for each example, so I cannot provide a reliable number here.

The size increase seems to result from the `block_on` function. When I remove the `block_on` calls from the `panic_handler` and `alloc_error_handler`, the binary size decreases significantly.

For further discussions, I would give rise to the following questions:

- What does the `text` section say about the performance of the code? From my understanding, RAM/speed optimized code is more inlined or monomorphized resulting in bigger binary size.
- What is the effect of `block_on` lying in a separate crate? `block_on` depends on `yieldk` which is bound at link time.

**Comment by jrvanwhy** [comment]
> > The downside of polling everything all the time is that each callback will result in work proportional to the total application size. I have no idea whether that could be a significant limitation. It also means that you an app developer needs to bring all futures back to a single point (i.e. `main()`) or we need a list of tasks to poll.
> 
> The current situation of `libtock-rs` is that you do _not_ poll everything. You only poll _the_ future of the `main` function. Considering that Tock processes are single-threaded and assuming that yielding within callbacks is bad, a single main future seems sufficient. Thus, there is no need for a waker keeping track which future has been awoken. To prevent multiple futures being spawned `block_on` is defined to be `unsafe`.

The size of that future is proportional to the total application size. Because we avoid using the wake mechanism, the future returned by main() will need to poll all active sub-futures each time.

> Regarding the point you mentioned about each callback resulting in some work: This has always been the case when using `yieldk_for`.

That's incorrect. With the previous synchronous API using `yieldk_for`, each callback does a minimal amount of work which does not scale with the size of the application. The worst case behavior for a single invocation of `yieldk_for` is running a number of callbacks that grows with the size of the application, but in the average case it averages out to 1 callback per `yieldk_for` invocation.

**Comment by jrvanwhy** [comment]
> * What does the `text` section say about the performance of the code? From my understanding, RAM/speed optimized code is more inlined or monomorphized resulting in bigger binary size.

The two are othogonal. Some speed optimizations shrink the binary size, some increase it. Some RAM optimizations shrink the binary's size, some improve it. On highly superscalar microarchitectures with large caches (e.g. desktop x86 chips), there is often a correlation between RAM usage and speed, but that is not true for the architectures Tock runs on. Also, the correlation doesn't run the other way: a large or small binary size gives no indication of runtime performance.

> * What is the effect of `block_on` lying in a separate crate? `block_on` depends on `yieldk` which is bound at link time.

It is monomorphized in the client's code, and as such for optimization purposes it is part of the end application.

**Comment by Woyten** [comment]
> The size of that future is proportional to the total application size.

I am not sure what you mean by *application size*. Clearly, the future cannot be proportional to the binary size as it is part of the binary.

From my understanding, the size of the future correlates with the size of the state that needs to be transported between yield points, i.o.w. between `await` calls. If you yielded manually using `yieldk`, how would this reduce the space required on the stack compared to the future?

> Because we avoid using the wake mechanism, the future returned by main() will need to poll all active sub-futures each time.

What do you mean by sub-future? Are you talking about concurrent futures using `join`? If yes, you are right: Each sub-future will be polled. This, however, is how the `join` method works and cannot be influenced by using a dedicated `Waker`.

Besides that, what's the cost of checking conditions of kind `cell.get() == my_expected_state`? This seems rather cheap compared to determining which future to invoke for which `Waker`.

Anyway, how is this comparable to a non-future-based API? We are talking about a `join` function which does not have an equivalent in the non-future world. I would propose not to focus too much on the `join` use case. The more common case, namely sequential `await`s, has no overhead with respect to polling: Only one future will polled at a time.

> That's incorrect. With the previous synchronous API using `yieldk_for`, each callback does a minimal amount of work which does not scale with the size of the application.

What I wanted to say is that the closure in `yieldk_for` was triggered no matter what caused the interruption. If you had 100 active buttons and a call `yieldk_for(|| button_0_pressed.get())` then the closure was getting called no matter which button was pressed. So, the scaling with the number of callbacks isn't new.

To sum it up, from my understanding, there are two numbers that the frequency of callback invocations scales with:
- Number of active interrupts (as without futures)
- Number of concurrent futures (only possible *with* futures and, therefore, not comparable)

**Comment by jrvanwhy** [comment]
> > The size of that future is proportional to the total application size.
> 
> I am not sure what you mean by _application size_. Clearly, the future cannot be proportional to the binary size as it is part of the binary.

By _application size_, I'm approximately describing the amount of concurrent tasks, code, and memory used by the app.

> From my understanding, the size of the future correlates with the size of the state that needs to be transported between yield points, i.o.w. between `await` calls. If you yielded manually using `yieldk`, how would this reduce the space required on the stack compared to the future?

I'm not referring to the size of the future in RAM, but rather to the amount of functionality contained in the future.

> > Because we avoid using the wake mechanism, the future returned by main() will need to poll all active sub-futures each time.
> 
> What do you mean by sub-future? Are you talking about concurrent futures using `join`? If yes, you are right: Each sub-future will be polled. This, however, is how the `join` method works and cannot be influenced by using a dedicated `Waker`.

Yes. Under a previous futures design (before 0.3), you could use a `Waker` to route wakeups to the correct sub-future. This evidently changed in futures 0.3.

> Besides that, what's the cost of checking conditions of kind `cell.get() == my_expected_state`? This seems rather cheap compared to determining which future to invoke for which `Waker`.
> 
> Anyway, how is this comparable to a non-future-based API? We are talking about a `join` function which does not have an equivalent in the non-future world. I would propose not to focus too much on the `join` use case. The more common case, namely sequential `await`s, has no overhead with respect to polling: Only one future will polled at a time.
> 
> > That's incorrect. With the previous synchronous API using `yieldk_for`, each callback does a minimal amount of work which does not scale with the size of the application.
> 
> What I wanted to say is that the closure in `yieldk_for` was triggered no matter what caused the interruption. If you had 100 active buttons and a call `yieldk_for(|| button_0_pressed.get())` then the closure was getting called no matter which button was pressed. So, the scaling with the number of callbacks isn't new.

In a synchronous API, the number of `yieldk_for` calls should be proportional to (and roughly equal to) the number of callbacks. The problem with the futures design is the number of callbacks remains approximately the same but the amount of work done per callback grows with the number of concurrent operations. An async design similar to what the kernel uses internally doesn't have the same complexity; each callback is O(1).

> To sum it up, from my understanding, there are two numbers that the frequency of callback invocations scales with:
> 
> * Number of active interrupts (as without futures)
> * Number of concurrent futures (only possible _with_ futures and, therefore, not comparable)

It is comparable to the concurrent-interacting-state-machines mechanism the Tock kernel uses.

File: size_comparison/README.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
+# Futures versus no-Futures Size Comparison
+
+Original author: Johnathan Van Why
+
```
Comment (phil-levis) [inline size_comparison/README.md:4]: Can you say what commit these numbers were generated for? That will let us keep track of the results.

**Comment by phil-levis** [comment]
I'd like to separate out review of @jrvanwhy 's numbers from their interpretation. In particular, it seems like nobody has comments on 

https://github.com/tock/design-explorations/blob/c38bf5dc47d4854eab8f735cb3295172185e7a6f/size_comparison/README.md

If we agree with his experimental results on the current code, then we should merge this. We can separately discuss the implications to libtock-rs and futures. Future work may change (improve or degrade) these numbers.

**Review by phil-levis — COMMENTED** [review]
Please add the commit hash, then we are good.

File: size_comparison/README.md (hunk 3 above)
Comment (jrvanwhy) [reply size_comparison/README.md:4]: I added a commit hash. The numbers were actually slightly out of date -- while fixing an entry point bug I moved 1 word from .text into .rt_header. The write-up and disassembly now match the latest commit.
